WIP: initial base draft for default extras (#1)
RonnyPfannschmidt/python-peps: pep-9990.rst

PEP: ?
Title: Specification of default extras in package metadata
Author: Ronny Pfannschmidt <opensource@ronnypfannschmidt.de

> Sponsor:

Reviewer: I'm happy to sponsor this as a PEP editor, though I'm not a packaging expert. The sponsor role is primarily about helping you through the PEP process, as noted in https://peps.python.org/pep-0001/#submitting-a-pep

```suggestion
Sponsor: Adam Turner <python@quite.org.uk>
```

Discussions-To:
Status: draft
Type: Standards Track
Topic: Packaging
Content-Type: text/x-rst
Created: 11-May-2023

Abstract
========

This PEP specifies how to declare names of optional dependencies that should be installed by default
and how to opt out of them

Motivation
============

> A number of python packages either ship requirements that are not strictly or they recommend installing with extras for certain use-caes.

Reviewer: ```suggestion
A number of python packages either ship requirements that are not strictly required or they recommend installing with extras for certain use-caes.
```

> For example, currently setuptools_scm_ has to pull in the minimal version of setuptools or risk confused users,
> even if tools like hatch_vcs_ or others can use it without needing setuptools.

Reviewer: ```suggestion
For example, currently setuptools_scm_ requires setuptools (to convey the minimum-supported setuptools), but it does not require setuptools at runtime.
`Hatch <hatch_vcs_>`_ would benefit from being able to specify setuptools_scm_, *without* having setuptools installed into the environment.
```

Additionally tools

> .. __setuptools_scm: https://github.com/pypa/setuptools_scm
> .. __hatch_vcs: https://github.com/ofek/hatch-vcs

Reviewer: ```suggestion
.. _setuptools_scm: https://github.com/pypa/setuptools_scm
.. _hatch_vcs: https://github.com/ofek/hatch-vcs
```

A ``Default-Extra: name`` metadata field should be added to list the extras that ought to be considered by default

recommended-dependencies in pyproject.toml
--------------------------------------------

.. code-block:: toml

   [project]
   name = "example"
   [project.recommended-dependencies]
   cli = [
      "click>=8.0",
   }

extras in requirement specifiers
---------------------------------

``-extra`` should be used to opt out of a specific extra

``-`` should be used as marker to opt out of all default extras

canonical extras requirement specifier

when a specific distribution is resolved/considered,
its extras specifier should be changed from a implicit form to a explicit form

Constraints
============

Rejected Ideas
==============

propagation of opt-outs
-----------------------

A requested opt-out of a default extra cannot safely be propagated,
as the effect on sibling dependencies is undefined.

To avoid ambiguity a resolver must transform the implied defaults into a explicit specification

References
===========

Copyright
=========

This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.

> 
> 
> ..
>    Local Variables:
>    mode: indented-text
>    indent-tabs-mode: nil
>    sentence-end-double-space: t
>    fill-column: 70
>    coding: utf-8
>    End:

Reviewer: We don't require this footer block anymore
```suggestion
```
